src/app/shared/wysiwyg/wysiwyg.component.ts

```diff
@@ -0,0 +1,392 @@
+/// <reference path="./prosemirror.d.ts"/>
+
+import { Component, Input, OnInit, OnChanges, OnDestroy, SimpleChanges, ElementRef, ViewChild } from '@angular/core';
+import { wrapIn, setBlockType, chainCommands, newlineInCode, toggleMark, baseKeymap } from 'prosemirror-commands';
+import { blockQuoteRule, orderedListRule, bulletListRule, codeBlockRule, headingRule,
+  inputRules,allInputRules } from 'prosemirror-inputrules';
+import { keymap } from 'prosemirror-keymap';
+import { schema, defaultMarkdownParser, defaultMarkdownSerializer } from 'prosemirror-markdown';
+import { MenuBarEditorView, toggleMarkItem, icons,
+  MenuItem, Dropdown, DropdownSubmenu,
+  wrapItem, blockTypeItem, joinUpItem, liftItem,
+  selectParentNodeItem} from 'prosemirror-menu';
+import { wrapInList, splitListItem, liftListItem, sinkListItem } from 'prosemirror-schema-list';
+import { EditorState, Plugin } from 'prosemirror-state';
+import { BaseModel } from '../model/base.model';
+import { ImageModel } from '../model/image.model';
+import { PromptComponent } from './prompt.component';
+
+@Component({
+  selector: 'publish-wysiwyg',
+  template: `
+    <div #contentEditable></div>
+    <publish-prompt #prompt>
+      <h1 class="modal-header">Link to</h1>
+      <div class="modal-body">
+        <label>URL<span class="error" [ngStyle]="{display: isURLInvalid() ? 'inline' : 'none'}">*</span></label>
+        <input type="text" name="url" #url [(ngModel)]="linkURL" required/>
+        <label>Title</label>
+        <input type="text" name="title" [(ngModel)]="linkTitle"/>
+        <p class="error" [ngStyle]="{display: isURLInvalid() ? 'block' : 'none'}">URL is required</p>
+      </div>
+      <div class="modal-footer">
+        <button (click)="createLink()">Okay</button>
+        <button (click)="prompt.hide()">Cancel</button>
+      </div>
+    </publish-prompt>
+  `
+})
+
+export class WysiwygComponent implements OnInit, OnChanges, OnDestroy {
```

Comment:
Holy state management!  That's a whole ton of stuff to handle for prosemirror!

To trim down this class a bit ... I'm wondering if some of the more image-y / state-management-y / keymap-y stuff could be moved to external classes/modules.  Something more "mixin" like, but could be accomplished with basic classes that you pass a few parameters to.  (Since we don't actually have mixins).